src/cloud-sql-instance.ts

```diff
@@ -98,60 +107,153 @@ export class CloudSQLInstance {
   async forceRefresh(): Promise<void> {
     // if a refresh is already ongoing, just await for its promise to fulfill
     // so that a new instance info is available before reconnecting
-    if (this.ongoingRefreshPromise) {
-      await this.ongoingRefreshPromise;
+    if (this.next) {
+      await this.next;
       return;
     }
     this.cancelRefresh();
     return this.refresh();
   }
 
   async refresh(): Promise<void> {
+    const currentRefreshId = this.scheduledRefreshID;
+
     // Since forceRefresh might be invoked during an ongoing refresh
     // we keep track of the ongoing promise in order to be able to await
     // for it in the forceRefresh method.
     // In case the throttle mechanism is already initialized, we add the
     // extra wait time `limitRateInterval` in order to limit the rate of
     // requests to Cloud SQL Admin APIs.
-    this.ongoingRefreshPromise = this.throttle
-      ? this.throttle(this._refresh).call(this)
-      : this._refresh();
-
-    // awaits for the ongoing promise to resolve, since the refresh is
-    // completed once the promise is resolved, we just free up the reference
-    // to the promise at this point, ensuring any new call to `forceRefresh`
-    // is able to trigger a new refresh
-    await this.ongoingRefreshPromise;
-    this.ongoingRefreshPromise = undefined;
-
-    // Initializing the rate limiter at the end of the function so that the
-    // first refresh cycle is never rate-limited, ensuring there are 2 calls
-    // allowed prior to waiting a throttle interval.
+    this.next = (
+      this.throttle && this.scheduledRefreshID
+        ? this.throttle(this.performRefresh).call(this)
+        : this.performRefresh()
+    )
+      // These needs to be part of the chain of promise referenced in
+      // next in order to avoid race conditions
+      .then((nextValues: RefreshResult) => {
+        // in case the id at the moment of starting this refresh cycle has
+        // changed, that means that it has been canceled
+        if (currentRefreshId !== this.scheduledRefreshID) {
+          return;
+        }
+
+        // In case the performRefresh method succeeded
+        // then we go ahead and update values
+        this.updateValues(nextValues);
+
+        this.scheduleRefresh();
```

Comment:
Could we pass in the next scheduled time here? As it, this looks like an immediate refresh.

Reply:
Having `scheduleRefresh()` be responsible of retrieving the scheduled time helps with not having to repeat this bit from L238 elsewhere:

```js
    const refreshInterval = getRefreshInterval(
      String(this.ephemeralCert?.expirationTime)
    );
```

Reply:
In all of our other connectors, we calculate the time and then pass it into schedule refresh -- this helps distinguish between an immediate refresh (`scheduleRefresh(0)`) and an eventual refresh `scheduleRefresh(refreshInterval)`.

Reply:
I see, I think @jackwotherspoon spotted a similar thing [here](https://github.com/GoogleCloudPlatform/cloud-sql-nodejs-connector/pull/219#discussion_r1337786834). Is `scheduleRefresh(0)` meant to be used for the fail path? and the `refreshInterval` on success?

Reply:
looks like in Go and Python we refresh immediately in the fail path: https://github.com/GoogleCloudPlatform/cloud-sql-go-connector/blob/4a28a788a94d64e1ce6ddd76fa3a041c82c8f2b1/internal/cloudsql/instance.go#L318-L355

Reply:
I'll update with that!